proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,477 @@
+- the status stored in the program account is `Deployed` or `Finalized`
+- the program account was not deployed within the current slot (delay
+visibility)
+- the executable file stored in the program account passes executable
+verification
+
+failing any of the above checks must throw `UnsupportedProgramId`.
+
+### Program Management Instructions
+
+The loader-v4 intructions Deploy and Retract are not authorized in CPI.
+
+#### Write
+
+- Instruction accounts:
+  - `[writable]` The program account to write to.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `0u32`
+  - `u32` Byte offset at which to write the given bytes
+  - `[u8]` Chunk of the programs executable file
+- Behavior:
+  - Charge 32 + chunk_length_in_bytes / cpi_bytes_per_unit CUs
+  - Check there are at least two instruction accounts,
+  otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header
+  - Check the status stored in the program account is `NeverBeenDeployed` or
+  `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Check that the end offset (sum of offset and length of the chunk) does
+  not exceed the maximum (program account length minus the header size),
+  otherwise throw `AccountDataTooSmall`
+  - Copy the chunk into the program account at the offset shifted by the
+  header size
+
+#### Copy
+
+- Instruction accounts:
+  - `[writable]` The program account to copy to.
+  - `[signer]` The authority of the program.
+  - `[]` The account to copy from.
+- Instruction data:
+  - Enum variant `1u32`
+  - `u32` Byte offset at which to write
+  - `u32` Byte offset at which to read
+  - `u32` Length of the chunk to copy in bytes
+- Behavior:
+  - Charge 32 + chunk_length_in_bytes / cpi_bytes_per_unit CUs
+  - Check there are at least three instruction accounts,
+  otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and source account do not alias,
+  otherwise throw `AccountBorrowFailed`
+  - Verify the program account header
+  - Check the status stored in the program account is `NeverBeenDeployed` or
+  `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Check that the source accounts owner and look-up the source header size:
+    - loader-v2: 0 bytes
+    - loader-v3 buffer: 37 bytes
+    - loader-v3 programdata: 45 bytes
+    - loader-v4: 48 bytes
+    - if none of the above matches throw `InvalidArgument`
+  - Check that the source end offset (sum of source offset and length) does
+  not exceed the maximum (source account length minus the source header size),
+  otherwise throw `AccountDataTooSmall`
+  - Check that the destination end offset (sum of destination offset and
+  length) does not exceed the maximum (program account length minus the loader-v4
+  header size), otherwise throw `AccountDataTooSmall`
+  - Copy the chunk between the program accounts at the offsets, each shifted by
+  the header size of their loader (account owner) respectively
+
+#### SetProgramLength
```

Comment:
If we stored the length of the program data in the program header, this instruction could be run on programs without retracting first. This would make program upgrades a little easier for smart contract controlled wallets because the instruction could be permissionless (as is the case in loader v3) and therefore not need to be done via CPI. I think this would be a big ergonomics improvement

Reply:
The primary benefit is that it can be called without CPI if the upgrade authority is a PDA